custom_components/iec/coordinator.py

```diff
@@ -247,10 +261,16 @@ async def _insert_statistics(self, contract_id: int, is_smart_meter: bool) -> No
             return
 
         _LOGGER.debug(f"Updating statistics for IEC Contract {contract_id}")
-        devices = await self.api.get_devices(str(contract_id))
-        month_ago_time = (datetime.now() - timedelta(weeks=4))
+        if contract_id in self._devices_by_contract_id:
+            devices = self._devices_by_contract_id[contract_id]
+        else:
+            devices = await self.api.get_devices(str(contract_id))
+            self._devices_by_contract_id[contract_id] = devices
```

Comment:
Repeated code, I suggest extracting it to function (will be easier to fix issue when needed in one place only)